On opt-out page move js to separate file to fix inline-script (CSP)

Skywalker-11 opened the pull request:
This PR moves the javascript part of the opt-out page that can be loaded via an iframe to a separate file. This allows the usage of CSP without the need to allow inline scripts or using hashes

Skywalker-11 committed:
on opt-out page move js to separate file to fix inline-script (CSP)

tsteur commented:
FYI: Merging this PR would break https://github.com/matomo-org/tracker-proxy/pull/37 which we finished yesterday cc @diosmosis

Skywalker-11 committed:
Compatibility of separate optOutJs with tracker-proxy

Skywalker-11 commented:
This PR should now also work with the patched tracker-proxy which PR is referenced above. To accieve this an additional get parameter (`matomoproxy`) is added on the proxy. If it doesn't exist in the request of the opt-out iframe the normal path `plugins/CoreAdminHome/javascripts/optOut.js` is used. 
If the parameter `matomoproxy` exists it will use `?module=CoreAdminHome&action=getOptOutJs` as path for the script which is translated by the proxy to the normal path which then will be returned.

I'm not that familiar with the matomo routing so if there is a better way to do this please tell me.

diosmosis commented:
I don't think using a special URL that doesn't actually work in matomo (ie, the `?module=CoreAdminHome&action=getOptOutJs`) as a sort of message passing between matomo and the tracker-proxy is a good idea. It's not really clear from looking at the matomo code that this is what that special URL is for and couples two repos together. This should ideally be handled entirely in the tracker-proxy (so matomo doesn't have to know the tracker-proxy exists). Something like, in the tracker-proxy for matomo-proxy.php, rewriting URLs to JS/CSS files in the output to point to matomo-proxy.php instead (eg, `tracker-proxy/matomo-proxy.php?file=plugins/CoreAdminHome/javascripts/optOut.js`). Would probably need to add a new global variable for this that is set in matomo-proxy.php & handled in proxy.php.

Skywalker-11 committed:
remove destinction between proxy and normal request

Skywalker-11 commented:
@mattab  Can you take an other look at the branches. I changed the proxy.php to replace the path in the content to point to the matomo-proxy.php?file<path>.
https://github.com/Skywalker-11/tracker-proxy/commits/master 
https://github.com/Skywalker-11/matomo/commits/optOutInlineScript

If that is ok with you I would create a new PR for the proxy and master

tsteur commented:
Was this closed by accident @mattab ?

diosmosis commented on misc/user/.htaccess:
```diff
@@ -0,0 +1,51 @@
+		Require all granted
+	</IfModule>
+</IfModule>
+</Files>
```
This file was added by accident I think.

diosmosis reviewed (commented):
Thanks for updating @Skywalker-11! There's an extra file that was added to the PR and tmp/.gitkeep shouldn't have it's file mode changed. I'll review the tracker-proxy changes next and we'll merge that first when everything's ready.

diosmosis commented:
BTW, if you could create a PR for your tracker-proxy changes (which in general look good), that would be helpful.

Skywalker-11 commented:
Yes files shouldn't be there. Maybe leftovers from setup or phpunit.
PR 42 of matomo-proxy referenced above should contain that changes. Do you want to reopen that PR or shall I create a new one?

diosmosis commented:
I'll reopen it.

Skywalker-11 committed:
revert unwanted changes of /tmp/.gitkeep and created .htaccess

diosmosis committed:
In optout form, clear new window closed check interval after new window is closed.

Otherwise, if reload takes more than 1s, the interval will run again and try another reload, cancelling the pending one. Which results in no reload occuring.
